base_layer/core/src/base_node/states/horizon_state_sync/state_sync.rs

```diff
@@ -0,0 +1,705 @@
+                if attempt == config.max_sync_request_retry_attempts {
+                    return Err(HorizonSyncError::MaxSyncAttemptsReached);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    // Synchronize UTXO MMR Nodes, RangeProof MMR Nodes and the UTXO set upto the horizon sync height from remote sync
+    // peers.
+    async fn synchronize_utxos_and_rangeproofs(&mut self) -> Result<(), HorizonSyncError> {
+        let config = self.shared.config.horizon_sync_config;
+        let local_num_utxo_nodes =
+            async_db::fetch_mmr_node_count(self.shared.db.clone(), MmrTree::Utxo, self.horizon_sync_height).await?;
+        let (remote_num_utxo_nodes, _sync_peer) = helpers::request_mmr_node_count(
+            LOG_TARGET,
+            self.shared,
+            self.sync_peers,
+            MmrTree::Utxo,
+            self.horizon_sync_height,
+            config.max_mmr_node_request_retry_attempts,
+        )
+        .await?;
+
+        for indices in (local_num_utxo_nodes..remote_num_utxo_nodes)
+            .collect::<Vec<u32>>()
+            .chunks(config.mmr_node_or_utxo_request_size)
+        {
+            for attempt in 1..=config.max_sync_request_retry_attempts {
+                let pos = indices.first().cloned().unwrap_or(0);
+                let count = indices.len() as u32;
+                let (utxo_hashes, utxo_bitmap, sync_peer1) = helpers::request_mmr_nodes(
+                    LOG_TARGET,
+                    self.shared,
+                    self.sync_peers,
+                    MmrTree::Utxo,
+                    pos,
+                    count,
+                    self.horizon_sync_height,
+                    config.max_mmr_node_request_retry_attempts,
+                )
+                .await?;
+                let (rp_hashes, _, sync_peer2) = helpers::request_mmr_nodes(
+                    LOG_TARGET,
+                    self.shared,
+                    self.sync_peers,
+                    MmrTree::RangeProof,
+                    pos,
+                    count,
+                    self.horizon_sync_height,
+                    config.max_mmr_node_request_retry_attempts,
+                )
+                .await?;
+
+                // Construct the list of hashes of the UTXOs that need to be requested.
+                let mut request_utxo_hashes = Vec::new();
+                let mut request_rp_hashes = Vec::new();
+                let mut is_stxos = Vec::with_capacity(utxo_hashes.len());
+                for index in 0..utxo_hashes.len() {
+                    let deleted = utxo_bitmap.contains(index as u32 + 1);
+                    is_stxos.push(deleted);
+                    if !deleted {
+                        request_utxo_hashes.push(utxo_hashes[index].clone());
+                        request_rp_hashes.push(rp_hashes[index].clone());
+                    }
+                }
+                // Download a partial UTXO set
```

Comment:
we can check the mmr hashses here as well, before we start downloading

Reply:
Ok - next PR?